.github/workflows/kind.yml

```diff
@@ -204,32 +205,32 @@ jobs:
       - name: Run e2e tests
         run: |
           mkdir log
-          mkdir test-e2e-encap-proxy-all-coverage
-          ANTREA_LOG_DIR=$PWD/log ANTREA_COV_DIR=$PWD/test-e2e-encap-proxy-all-coverage ./ci/kind/test-e2e-kind.sh --encap-mode encap --proxy-all --coverage --skip mode-irrelevant
+          mkdir test-e2e-encap-all-features-disabled-coverage
+          ANTREA_LOG_DIR=$PWD/log ANTREA_COV_DIR=$PWD/test-e2e-encap-all-features-disabled-coverage ./ci/kind/test-e2e-kind.sh --encap-mode encap --coverage --feature-gates AllAlpha=false,AllBeta=false
```

Comment:
Should we just test with default values - keep default enabled features enabled, and default disabled features disabled? Usually, no one will disable default enabled features right?

Reply:
"test-e2e-encap" tests the configuration you describe, it has been there from the begining.
If we don't have this job "all-features-disabled", there will be no job testing the code under the `! DefaultFeatureGate.Enabled(FEATURE)` condition, which were previously covered by "test-e2e-encap-no-proxy", "test-e2e-encap-no-np". Is it safe that once a feature graduates to beta, we never run test with it being disabled?

Reply:
@jianjuns do you agree with above or have other comments?

Reply:
Then the feature specific code under `! DefaultFeatureGate.Enabled(FEATURE)` except AntreaProxy will not be covered by any e2e test, is this reliable? For example, if a PR makes a change to a flow in L3Forwarding with the assumption that the packet will be handled by some flows installed by Egress feature, no test would catch the problem. It's not normal that users disable features that are enabled by default, but should we cover them in test with best effort? this may also affect  test coverage.

Reply:
My assumption is that AntreaProxy or not may impact the pipeline and other features more. If we believe disabling AntreaProxy is indeed required by some users, we should test it with other options to be default values.

We anyway cannot cover all feature gate combinations, and so there are always risks some combination can be broken. If we believe both "all features disabled" and "only Antrea Proxy are important too, seems we need two tests for them respectively.

Reply:
I added "all features disabled" not because it's an important combination, but a single job to cover the code path that all optional features are disabled and ensure core functions work properly without relying on any optional feature (When these features are disabled, no feature specific test case will run, the test cases are only the generic ones like Pod connectivey, K8s NetworkPolicy, etc.). I think it's not just adding 1 feature combination, but 0 coverage vs. full coverage on related code path, "all features enabled" covers the other side. And I think "all features disabled" covers "no-proxy" in some way from core functions's perspective, assuming optional features are not too coupled with each other, (and if they are coupled explicitly, disabling either of them is not a supported scenario and we won't even test it).

Reply:
Regarding configuration combination, I very agree we cannot cover all combinations, I think all tests are only testing the default combination with single factor mutated:
test-e2e-ipv6 and test-e2e-ipv6-only's factor is ip family
test-windows's factor is platform
test-e2e-noencap and test-e2e-hybrid's factor is encapulation mode
test-e2e-all-features-enabled and test-e2e-all-features-disabled is planed for the factor of feature

Reply:
I think the questions are:

1. do we believe "AntreaProxy disabled" is a case used by considerable number of users
2. if so, are these users would like other default features enabled
3. if so, could this combination of "AntreaProxy disabled" + "other features enabled" have a good chance to be broken, if we do not test it

Do you mean you think the chance of 3) is low, if we test "AntreaProxy and all other features disabled"? I feel AntreaProxy flows can impact other features like ANP and Egress, no?

Reply:
1. I only know one company disables it because of insufficient kernel version
2. I guess no because the above reason doesn't apply to other features, however, just disabling AntreaProxy and keeping other configuration as default is not exactly their configuration. AFAIK, at least they change encapsulation
mode and tunnel type: they use both noEncap mode and encap mode with vxlan tunnel. I'm not sure other configurations. 
3. It has a chance to be broken, but I imagine it shouldn't be too related to other features. For examples, ANP doesn't care whether it's kube proxy or antrea proxy, it cares the 5-tuple anyway, with one exception, it relies on AntreaProxy to implement `toServices`, but it's documented as unsupported and test didn't run when AntreaProxy is disabled. For Egress, if the traffic is desined to external address directly, how proxy works doesn't matter at all; if the traffic is destined to Service whose backend is external address, I guess it even doesn't work today when AntreaProxy is disabled because we match `InPort` when determining whether to SNAT.

I agree with your proposal after reconsidering the test from another perspective: every feature should finally graduate to GA or deprecated, however as long as the kernel version restriction exists or users have a reason to stick to kube-proxy, AntreaProxy will GA and be removed from feature gate and perhaps a configuration "antreaProxy.enable" should be added like `--run-service-proxy` of kube-router. And if we think it's an important scenario like noEncap, we could keep a dedicated job testing it being disabled and all features being kept as default. Before that happens, we configure the job via "AntreaProxy feature gate", but it's for a typical configuration like noEncap. From feature gate's perspective, we only test a job with their default setting and a job with all of them enabled, the former of which covers the most common configuration and the latter ensures a non-default feature will be tested.

Reply:
Yes, I always wonder if we really need to test with default enabled feature gates disabled. The only reason to disable a feature gate should be either defects or maybe a feature does not work in particular envs. We should know if we really have the latter case (and ideally feature gate should not be for that).